### PR TITLE
[core] Replace subprocess.Popen with safer get_subprocess_output

### PR DESCRIPTION
**Pull request by JohnLZeller**
Issue came up when investigating hostdown issues.
https://datadog.zendesk.com/agent/tickets/30510

It seems as though, periodically, a subprocess call hangs and ultimately causes the watchdog to reset the agent. During this period of time, there is then a host down event, followed by a host recovered event posted to the customers events page. We noticed that it seems as though the hanging happens in various checks (not the same for every host down event). Offending checks have included the customers own custom `iostat` check, as well as our own `postfix` check, and even `resources/processes.py`.

The theory is that this is caused by a hanging call to `subprocess.Popen`. There is a known issue where the use of a `subprocess.PIPE` for stdout (or even stderr) could cause the process to hang once the `subprocess.PIPE` fills up. The solution to this is to use `get_subprocess_output()` in our own `utils/subprocess_output.py`, which uses a file for stdout and stderr, removing such a low memory capacity issue.

To avoid further chasing this problem around the agent, this PR attempts to replace the use of `subprocess.Popen` anywhere it shows up in the agent with `get_subprocess_output()`. There are a few instances where further work will be needed to refactor the code in a clean way, and for those instances, I have added a `FIXME` note.

### CODE REVIEW

**Comment by JohnLZeller** [comment]
I have tested these on personal_chef, and fixed any issues that came up.

File: checks.d/postfix.py [hunk 1]
```diff
@@ -55,8 +57,13 @@ def _get_queue_count(self, directory, queues, tags):
                 # can dd-agent user run sudo?
                 test_sudo = os.system('setsid sudo -l < /dev/null')
                 if test_sudo == 0:
-                    count = os.popen('sudo find %s -type f | wc -l' % queue_path)
-                    count = count.readlines()[0].strip()
+                    find = get_subprocess_output(['sudo', 'find', queue_path, '-type', 'f'], self.log)
```
Comment (remh) [inline checks.d/postfix.py]: the second command is
```
wc -l
```
so you can change this whole block with:
```python
count = len(get_subprocess_output(['sudo', 'find', queue_path, '-type', 'f'], self.log).splitlines())
```

**Comment by JohnLZeller** [comment]
Reference material: http://thraxil.org/users/anders/posts/2008/03/13/Subprocess-Hanging-PIPE-is-your-enemy/

File: checks.d/postfix.py (hunk 1 above)
Comment (JohnLZeller) [inline checks.d/postfix.py]: aha, clever :) I missed it! *shakes fist at the sky*

File: config.py [hunk 2]
```diff
@@ -581,15 +583,20 @@ def get_system_stats():
     platf = sys.platform
 
     if Platform.is_linux(platf):
-        grep = subprocess.Popen(['grep', 'model name', '/proc/cpuinfo'], stdout=subprocess.PIPE, close_fds=True)
-        wc = subprocess.Popen(['wc', '-l'], stdin=grep.stdout, stdout=subprocess.PIPE, close_fds=True)
-        systemStats['cpuCores'] = int(wc.communicate()[0])
+        grep = get_subprocess_output(['grep', 'model name', '/proc/cpuinfo'], log)
+        # Must use tempfiles to redirect stdout to second command
+        tempgrep = tempfile.TemporaryFile()
```
Comment (remh) [inline config.py]: same thing here

File: utils/subprocess_output.py [hunk 3]
```diff
@@ -8,16 +8,23 @@
...
     if an error occurs.
     """
+
+    # Use tempfile, allowing a larger amount of memory. The subprocess.Popen
+    # docs warn that the data read is buffered in memory. They suggest not to
+    # use subprocess.PIPE if the data size is large or unlimited.
     with tempfile.TemporaryFile('rw') as stdout_f:
-        proc = subprocess.Popen(command, close_fds=True,
-                                stdout=stdout_f, stderr=subprocess.PIPE)
+        stderr_f = tempfile.TemporaryFile('rw')
```
Comment (remh) [inline utils/subprocess_output.py]: with statements are your friend :)

**Comment by JohnLZeller** [comment]
@remh incorporated all your changes :)

Sorry I sorta jumped the :gun: on squashing the commits.
Here are the changes:
[postfix.py Line 59](https://github.com/DataDog/dd-agent/pull/1892/files#diff-7bb12651742be50bad442a75e07fd0f8R59)
[config.py Line 585](https://github.com/DataDog/dd-agent/pull/1892/files#diff-fdf313ed2303059c44a9ebfffbef8eebR585)
[subprocess_output.py Line 21](https://github.com/DataDog/dd-agent/pull/1892/files#diff-463543c96bdaf50254bcaf9eb154ed3eR21)

All changes have been tested! :+1:

File: checks.d/mysql.py [hunk 4]
```diff
@@ -385,8 +385,7 @@ def _get_server_pid(self, db):
         if pid is None:
             try:
                 if sys.platform.startswith("linux"):
-                    ps = subprocess.Popen(['ps', '-C', 'mysqld', '-o', 'pid'],
-                                          stdout=subprocess.PIPE, close_fds=True).communicate()[0]
+                    ps = get_subprocess_output(['ps', '-C', 'mysqld', '-o', 'pid'], self.log)
```
Comment (JohnLZeller) [inline checks.d/mysql.py]: Use splitlines()

File: utils/subprocess_output.py [hunk 5]
```diff
@@ -8,22 +8,29 @@
...
-                                stdout=stdout_f, stderr=subprocess.PIPE)
-        proc.wait()
-        err = proc.stderr.read()
-        if err:
-            log.debug("Error while running {0} : {1}".format(" ".join(command),
-                                                             err))
-
-        stdout_f.seek(0)
-        output = stdout_f.read()
+        with tempfile.TemporaryFile('rw') as stderr_f:
```
Comment (remh) [inline utils/subprocess_output.py]: http://stackoverflow.com/questions/3024925/python-create-a-with-block-on-several-context-managers

Comment (JohnLZeller) [inline utils/subprocess_output.py]: aha, I didn't find this post. thanks!

**Comment by remh** [comment]
@JohnLZeller can you fix the conflicts please ?

File: checks.d/varnish.py [hunk 6]
```diff
@@ -97,21 +98,13 @@ def check(self, instance):
...
-        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
-                                stderr=subprocess.PIPE)
-        output, error = proc.communicate()
-        if error and len(error) > 0:
-            self.log.error(error)
-        return output
-
     def _get_version_info(self, varnishstat_path):
         # Get the varnish version from varnishstat
+        # FIXME: Use get_subprocess_output() instead of subprocess.Popen
```
Comment (remh) [inline checks.d/varnish.py]: Can you do it while we're at it ?

File: checks/collector.py [hunk 7]
```diff
@@ -642,6 +642,7 @@ def _populate_payload_metadata(self, payload, check_statuses, start_event=True):
                     command = "gohai"
                 else:
                     command = "gohai\gohai.exe"
+                # FIXME: Use get_subprocess_output() instead of subprocess.Popen
```
Comment (remh) [inline checks/collector.py]: Same here

File: jmxfetch.py [hunk 8]
```diff
@@ -275,6 +275,7 @@ def _start(self, path_to_java, java_run_opts, jmx_checks, command, reporter, too
             log.info("Running %s" % " ".join(subprocess_args))
 
             # Launch JMXfetch subprocess
+            # FIXME: Use get_subprocess_output() instead of subprocess.Popen
```
Comment (remh) [inline jmxfetch.py]: Same here

Comment (JohnLZeller) [inline jmxfetch.py]: This one is gonna take some more work in the morning. This function returns the return code of the process, making the return of get_subprocess_output inadequate, unless the caller of this function is okay with either a 0 or 1 return code, representing either an error string that is either empty or not.

**Comment by JohnLZeller** [comment]
Fixed a couple of the FIXME's. Needs a little more love, and some testing.

**Comment by remh** [comment]
Thanks @JohnLZeller can you fix the conflicts please ?

**Comment by JohnLZeller** [comment]
Okay, fixed the other places where there were existing FIXME comments for subprocess, as well as merged master to catch this branch up, and resolved conflicts.

Had to make some changes to the way the `get_subprocess_output()` function works, in order to allow it to work for all the cases we needed it. I did not however end up using it for `jmxfetch.py` because it would have made `get_subprocess_output()` too messy. Instead I implemented a safer way of calling subprocess for `jmxfetch.py`.

**Comment by JohnLZeller** [comment]
Could use some feedback.

@olivielpeau would you mind reviewing half? And @yannmh could you review the other half?

File: checks.d/disk.py [hunk 9]
```diff
@@ -176,9 +176,7 @@ def collect_latency_metrics(self):
 
     # no psutil, let's use df
     def collect_metrics_manually(self):
-        df_out = utils.subprocess_output.get_subprocess_output(
-            self.DF_COMMAND + ['-k'], self.log
-        )
+        df_out, err, rtcode = get_subprocess_output(self.DF_COMMAND + ['-k'], self.log)
```
Comment (olivielpeau) [inline checks.d/disk.py]: Just a general comment when using function that return multiple values: you can discard the values that you don't use with `_`. It's just a convention but makes it clear that you're not using the value afterwards. For instance here you can use:

```python
df_out, _, _ = get_subprocess_output(self.DF_COMMAND + ['-k'], self.log)
```

File: checks.d/mysql.py [hunk 10]
```diff
@@ -385,9 +385,8 @@ def _get_server_pid(self, db):
         if pid is None:
             try:
                 if sys.platform.startswith("linux"):
-                    ps = subprocess.Popen(['ps', '-C', 'mysqld', '-o', 'pid'],
-                                          stdout=subprocess.PIPE, close_fds=True).communicate()[0]
-                    pslines = ps.strip().split('\n')
+                    ps, err, rtcode = get_subprocess_output(['ps', '-C', 'mysqld', '-o', 'pid'], self.log)
+                    pslines = ps.strip().splitlines()
                     # First line is header, second line is mysql pid
                     if len(pslines) == 2 and pslines[1] != '':
```
Comment (olivielpeau) [inline checks.d/mysql.py]: Nitpick: Now that we're using `splitlines()` instead of `split("\n")` we don't need to test `pslines[1] != ''` anymore, right? (as `splitlines()` removes the last line when it's empty) :)

File: utils/subprocess_output.py [hunk 11]
```diff
@@ -8,23 +9,29 @@
 
 
 # FIXME: python 2.7 has a far better way to do this
-def get_subprocess_output(command, log):
+def get_subprocess_output(command, log, shell=None, stdin=None):
```
Comment (olivielpeau) [inline utils/subprocess_output.py]: Probably has no impact at all, but the default value of `shell` in `subprocess.Popen` is `False`, not `None` (see https://docs.python.org/2/library/subprocess.html#subprocess.Popen).

File: jmxfetch.py [hunk 12]
```diff
@@ -274,26 +276,27 @@ def _start(self, path_to_java, java_run_opts, jmx_checks, command, reporter, too
...
-                sys.stdout.write(out)
-                sys.stderr.write(err)
-            else:
-                # Wait for JMXFetch to return
-                jmx_process.wait()
+            # Launch JMXfetch subprocess manually, w/o get_subprocess_output(), since it's a special case
+            with nested(tempfile.TemporaryFile('rw'), tempfile.TemporaryFile('rw')) as (stdout_f, stderr_f):
+                jmx_process = subprocess.Popen(
+                    subprocess_args,
+                    close_fds=not redirect_std_streams,  # set to True instead of False when the streams are redirected for WIN compatibility
```
Comment (olivielpeau) [inline jmxfetch.py]: Comes from one of my commits actually but the comment here is wrong, it's not `set to True when the streams are redirected` but rather `only set to True when the streams are not redirected`. Could you correct it please? :)

**Comment by olivielpeau** [comment]
Great job, looks way cleaner now!

Looks good to me apart from my comments :+1:

**Comment by JohnLZeller** [comment]
Thanks @olivielpeau :)

Made the changes you mentioned, thanks for the good feedback!

**Comment by remh** [comment]
@JohnLZeller can you squash your commits please ?

**Comment by JohnLZeller** [comment]
@remh will do

**Comment by JohnLZeller** [comment]
@remh okay, I have squashed and rebased to matched up with the current state of master

**Comment by miketheman** [comment]
I'm very interested in trying out some of these methods in a new check, where might I find a recent build that includes this code?

**Comment by miketheman** [comment]
Reason i ask is that http://apt.datadoghq.com/dists/nightly/ has datadog-agent_5.5.0.git.6.ff669c3-1_amd64.deb - which is [back on Sept 17](https://github.com/DataDog/dd-agent/commit/ff669c3).

**Comment by remh** [comment]
http://apt.datad0g.com/dists/nightly/

**Comment by miketheman** [comment]
Thanks @remh. I must have missed that when looking.
